schemachange/sc_queues.c

```diff
@@ -786,18 +786,147 @@ static int perform_trigger_update_int(struct schema_change_type *sc)
+	int len = 0;
+	char *schema_start = "schema {cstring type[4] cstring tbl[64] datetime logtime ";
+    len += strlen(schema_start);
+	char *line_postfix = "null=yes ";
+	for(int i = 0; i < schema->nmembers; i++){
+		struct field *entry = schema->member + i;
+		int line_size = strlen(csc2type(entry)) + 1 + strlen(entry->name) + strlen(line_postfix);
+        // zTODOc: their should be a list of types that can do this
+        if (entry->type == SERVER_BCSTR || entry->type == SERVER_BYTEARRAY){
+            // Types that have []
+            int len_size = floor(log(entry->len)) + 1;
+            line_size += 2 + len_size;
+        }
+		int new_line = line_size + 5; /* +5 is for the "new_ " */
+		int old_line = line_size + 5; /* +5 is for the "old_ " */
+		len += new_line + old_line;
+	}
+    /* "}" */
+    len += 1;
+	char *audit_schema = malloc((len + 1) * sizeof(char));
+	strcpy(audit_schema, schema_start);
+	for(int i = 0; i < schema->nmembers; i++){
+		struct field *entry = schema->member + i;
+        
+		char *type = csc2type(entry);
+		char *name = NULL;
+        
+        if (entry->type == SERVER_BCSTR || entry->type == SERVER_BYTEARRAY){
+            int len_size = floor(log(entry->len)) + 1;
+            char len[len_size + 3];
+            sprintf(len, "[%d]", entry->len);
+            name = malloc((strlen(entry->name) + strlen(len) + 1));
+            strcpy(name, entry->name);
+            strcat(name, len);
+        } else {
+            name = malloc((strlen(entry->name) + 1));
+            strcpy(name, entry->name);
+        }
+
+		strcat(audit_schema, type);
+		strcat(audit_schema, " new_");
+		strcat(audit_schema, name);
+		strcat(audit_schema, " ");
+		strcat(audit_schema, line_postfix);
+
+		strcat(audit_schema, type);
+		strcat(audit_schema , " old_");
+		strcat(audit_schema , name);
+		strcat(audit_schema , " ");
+		strcat(audit_schema , line_postfix);
+
+        free(name); 
+
+	}
+    
+	strcat(audit_schema, "}");
+	/* Assert that we got the length correct */
+    assert(strlen(audit_schema) == len);
+	return audit_schema;
+}
+
+
+struct schema_change_type *create_audit_table_sc(char *name){
+	struct schema_change_type *sc = new_schemachange_type();
+    sc->sc_chain_next = NULL;
+
+	// It's totally possible any of the immediatly below section is wrong
+    sc->onstack = 1;
+	sc->type = DBTYPE_TAGGED_TABLE;
+	sc->scanmode = gbl_default_sc_scanmode;
+    sc->live = 1;
+
+	sc->addonly = 1;
+
+	char *prefix = "$audit_";
+	strcpy(sc->tablename, prefix);
+	strcat(sc->tablename, name);
+	struct dbtable *db = get_dbtable_by_name(name);
+	sc->newcsc2 = get_audit_schema(db->schema);
+
+    if (db->instant_schema_change) sc->instant_sc = 1;
+
+	// What is ODH? This is just copied from timepart
+	if (db->odh) sc->headers = 1;
+
+    make_name_available(sc->tablename);
+
+	return sc;
+}
+
+char *malloced_strcpy(char *str){
+    char *new_str = malloc(strlen(str) + 1);
+    strcpy(new_str, str);
+    return new_str;
+}
```

Comment:
Try this instead: https://pubs.opengroup.org/onlinepubs/9699919799/functions/strdup.html

Reply:
Lol probably should've figured that was a built-in